### PR TITLE
STOIC21 dataset

### PR DESCRIPTION
**Pull request by mosheraboh**

### CODE REVIEW

**Comment by mosheraboh** [comment]
@taltlusty @liamhazan @SagiPolaczek @itaijj @egozi @alex-golts, here is another example of a dataset implementation (still wasn't reviewed or merged to fuse2 branch)
You are welcome to use it as a reference and/or review and add comments.

File: fuseimg/datasets/tests/test_datasets.py [hunk 1]
```diff
@@ -7,18 +7,35 @@
...
+        
         self.kits21_cache_dir = mkdtemp(prefix="kits21_cache")
         self.kits21_data_dir = mkdtemp(prefix="kits21_data")
-    def test_kits32(self):
+
+        self.stoic21_cache_dir = mkdtemp(prefix="stoic_cache")
+
+
+
+    def test_kits21(self):
```
Comment (SagiPolaczek) [inline fuseimg/datasets/tests/test_datasets.py:38]: Finally was changed from 32 to 21!
It bothered me lol :)

File: fuse/data/datasets/caching/samples_cacher.py [hunk 2]
```diff
@@ -344,7 +345,7 @@ def _cache(self, orig_sample_id:Any):
     
         
 
-def _get_available_write_location(cache_dirs:List[str], max_allowed_used_space=0.95):               
+def _get_available_write_location(cache_dirs:List[str], max_allowed_used_space=None):
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fuse/data/datasets/caching/samples_cacher.py:348]: I will change it back, I've changed it so I can keep working.

File: fuse/data/datasets/dataset_default.py [hunk 3]
```diff
@@ -96,7 +97,7 @@ def create(self, num_workers:int = 0) -> None:
         elif self._allow_uncached_sample_morphing:
             _output_sample_ids_info_list = run_multiprocessed(DatasetDefault._process_orig_sample_id, 
                 [(sid, self._static_pipeline, False) for sid in self._orig_sample_ids], 
-                workers=num_workers)    
+                workers=num_workers, mp_context=mp_context)
```
Comment (YoelShoshan) [inline fuse/data/datasets/dataset_default.py:100]: a suggestion - maybe multiprocessing_start_method is a better name ?

File: fuse/data/datasets/dataset_default.py [hunk 4]
```diff
@@ -199,10 +200,11 @@ def _getitem_multiprocess(item: Union[Hashable, int, np.integer]):
         return dataset.getitem(item, **kwargs)
 
 
-    def get_multi(self, items: Optional[Sequence[Union[int, Hashable]]] = None, workers: int = 10, verbose: int = 1, **kwargs) -> List[Dict]:
+    def get_multi(self, items: Optional[Sequence[Union[int, Hashable]]] = None, workers: int = 10, verbose: int = 1, mp_context: Optional[str] = None, **kwargs) -> List[Dict]:
```
Comment (YoelShoshan) [inline fuse/data/datasets/dataset_default.py:203]: multiprocessing_start_method ?

File: fuse/data/datasets/dataset_default.py [hunk 5]
```diff
@@ -199,10 +200,11 @@ def _getitem_multiprocess(item: Union[Hashable, int, np.integer]):
...
         """
         See super class
-        :param workers: number of processes to read the data. set to 0 for a single process.  
+        :param workers: number of processes to read the data. set to 0 for a single process.
```
Comment (YoelShoshan) [inline fuse/data/datasets/dataset_default.py:206]: -> "set to 0 to not use multi processing (useful when debugging)"

maybe a subtlety but when set to 1 you actually do create a single process that reads the data, which isn't the main process.

File: fuse/data/ops/op_base.py [hunk 6]
```diff
@@ -101,7 +101,7 @@ def op_call(op: OpBase, sample_dict: NDict, op_id: str, **kwargs):
             return op(sample_dict, **kwargs)
     except:
         # error messages are cryptic without this. For example, you can get "TypeError: __call__() got an unexpected keyword argument 'key_out_input'" , without any reference to the relevant op!
-        print(f'error in __call__ method of op={op}, op_id={op_id}, sample_id={get_sample_id(sample_dict)} - more details above')   
+        print(f'error in __call__ method of op={op}, op_id={op_id}, sample_id={get_sample_id(sample_dict)} - more details below')
```
Comment (YoelShoshan) [inline fuse/data/ops/op_base.py:104]: good catch! :)

File: fuse/data/ops/ops_common.py [hunk 7]
```diff
@@ -351,6 +353,46 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], keep_keypaths:List[
         sample_dict = NDict()
         for k in keep_keypaths:
             sample_dict[k] = prev_sample_dict[k]
+
+        return sample_dict
+
+
+class OpMap(OpBase):
```
Comment (YoelShoshan) [inline fuse/data/ops/ops_common.py:360]: just a thought - "map" became associated with running a set of elements through a provided function.

Mabye OpDictionaryLookup or OpLookup ? 

feel free to ignore if you don't agree.

File: fuse/data/ops/ops_common.py [hunk 8]
```diff
@@ -351,6 +353,46 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], keep_keypaths:List[
...
+    def __call__(self, sample_dict: NDict, key_in: str, key_out: str) -> Union[None, dict, List[dict]]:
+        value = sample_dict[key_in]
+        one_hot = np.zeros(self._num_classes)
+        one_hot[value] = 1.0
+        sample_dict[key_out] = one_hot
+
+        return sample_dict
+
+
+class OpConcat(OpBase):
```
Comment (YoelShoshan) [inline fuse/data/ops/ops_common.py:392]: can you please add a docstring and also an example of input and output for each Op class ?

File: fuse/utils/ndict.py [hunk 9]
```diff
@@ -59,22 +59,22 @@ class NDict(dict):
 
     """
     
-    def __init__(self, d: Union[dict, tuple, types.GeneratorType, NDict, None]=None):        
+    def __init__(self, dict_like: Union[dict, tuple, types.GeneratorType, NDict, None]=None):
```
Comment (YoelShoshan) [inline fuse/utils/ndict.py:62]: good catch! much better name :)

**Review by YoelShoshan — DISMISSED** [review]
LGTM
minor comments inline

**Pull request merged**